REST Client with generic subresource fails with InvalidRestClientDefinitionException

### Describe the bug

Given the following REST client definition:

```java
@Path("api")
@RegisterRestClient
public interface TestApi {

	interface GenericFunction<R> {

		@GET
		R get();
	}

	@Path("get")
	GenericFunction<String> subresource();
}
```

fails during invocation (not deployment, yikes) with:
```posh
Caused by: org.jboss.resteasy.reactive.client.api.InvalidRestClientDefinitionException: Failed to generate client for class interface org.acme.TestApi : Sub resource type is not a class: org.acme.TestApi$GenericFunction
        at org.jboss.resteasy.reactive.client.impl.ClientProxies.get(ClientProxies.java:29)
        at org.jboss.resteasy.reactive.client.impl.WebTargetImpl.proxy(WebTargetImpl.java:449)
        at io.quarkus.rest.client.reactive.runtime.RestClientBuilderImpl.build(RestClientBuilderImpl.java:386)
        ... 38 more
```

Due to the fact, that the actual type argument is a very concrete class, I suspect the extension the unresolved `GenericFunction#get` method of the class itself being checked (returning a type variable of course) instead of the actual parameterized type, which is returned by `TestApi#subresource`.

If I change the REST client definition to the following, everything works of course:
```java
@Path("api")
@RegisterRestClient
public interface TestApi {

	interface SimpleFunction {

		@GET
		String get();
	}

	@Path("get")
	SimpleFunction subresource();
}
```

### Expected behavior

As the given generic subresource is a valid subresource locator with a valid subresource method when implementing this API as an API provider, I would expect this API interface being a valid defintion for a REST client too. For this generic use case, all static information are available to resolve the the subresource method's type variable at deployment time to String, not some unbound type variable. The `quarkus-resteasy-reactive` extension is able to do so, but not the `quarkus-rest-client-reactive` extension.

### Actual behavior

Fails with the beforementioned exception during invocation time :/ 

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

Darwin sl-macbook.fritz.box 22.6.0 Darwin Kernel Version 22.6.0: Fri Sep 15 13:39:54 PDT 2023; root:xnu-8796.141.3.700.8~1/RELEASE_ARM64_T8112 arm64

### Output of `java -version`

openjdk version "17.0.8.1" 2023-08-24 OpenJDK Runtime Environment Homebrew (build 17.0.8.1+0) OpenJDK 64-Bit Server VM Homebrew (build 17.0.8.1+0, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.4.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6

### Additional information

_No response_

--- a/extensions/resteasy-reactive/rest-client-jaxrs/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java
+++ b/extensions/resteasy-reactive/rest-client-jaxrs/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java
@@ -5,6 +5,7 @@ import static org.jboss.jandex.Type.Kind.ARRAY;
 import static org.jboss.jandex.Type.Kind.CLASS;
 import static org.jboss.jandex.Type.Kind.PARAMETERIZED_TYPE;
 import static org.jboss.jandex.Type.Kind.PRIMITIVE;
+import static org.jboss.jandex.Type.Kind.TYPE_VARIABLE;
 import static org.jboss.resteasy.reactive.client.impl.RestClientRequestContext.DEFAULT_CONTENT_TYPE_PROP;
 import static org.jboss.resteasy.reactive.common.processor.EndpointIndexer.extractProducesConsumesValues;
 import static org.jboss.resteasy.reactive.common.processor.JandexUtil.isAssignableFrom;
@@ -83,6 +84,7 @@ import org.jboss.jandex.MethodParameterInfo;
 import org.jboss.jandex.ParameterizedType;
 import org.jboss.jandex.PrimitiveType;
 import org.jboss.jandex.Type;
+import org.jboss.jandex.TypeVariable;
 import org.jboss.logging.Logger;
 import org.jboss.resteasy.reactive.client.api.ClientMultipartForm;
 import org.jboss.resteasy.reactive.client.handlers.ClientObservabilityHandler;
@@ -930,7 +932,7 @@ public class JaxrsClientReactiveProcessor {
                     handleSubResourceMethod(enrichers, generatedClasses, interfaceClass, index, defaultMediaType,
                             httpAnnotationToMethod, name, classContext, baseTarget, methodIndex, method,
                             javaMethodParameters, jandexMethod, multipartResponseTypes, Collections.emptyList(),
-                            generatedSubResources);
+                            generatedSubResources, new HashMap<>());
                 } else {
                     FieldDescriptor methodField = classContext.createJavaMethodField(interfaceClass, jandexMethod,
                             methodIndex);
@@ -1228,7 +1230,7 @@ public class JaxrsClientReactiveProcessor {
                     handleReturn(interfaceClass, defaultMediaType, method.getHttpMethod(),
                             method.getConsumes(), jandexMethod, methodCreator, formParams,
                             bodyParameterIdx == null ? null : methodCreator.getMethodParam(bodyParameterIdx), builder,
-                            multipart);
+                            multipart, Collections.emptyMap());
                 }
             }
 
@@ -1399,9 +1401,48 @@ public class JaxrsClientReactiveProcessor {
             ClassRestClientContext ownerContext, ResultHandle ownerTarget, int methodIndex,
             ResourceMethod method, String[] javaMethodParameters, MethodInfo jandexMethod,
             Set<ClassInfo> multipartResponseTypes, List<SubResourceParameter> ownerSubResourceParameters,
-            Map<GeneratedSubResourceKey, String> generatedSubResources) {
+            Map<GeneratedSubResourceKey, String> generatedSubResources, Map<String, Type> ownerIdentifierToTypeVariable) {
+
+        Map<String, Type> identifierToTypeVariable = new HashMap<>();
         Type returnType = jandexMethod.returnType();
-        if (returnType.kind() != CLASS) {
+        if (returnType.kind() == PARAMETERIZED_TYPE) {
+
+            ParameterizedType parameterizedReturnType = returnType.asParameterizedType();
+            ClassInfo returnClass = index.getClassByName(returnType.name());
+            ParameterizedType.Builder methodReturnTypeBuilder = ParameterizedType.builder(returnType.name());
+            for (int i = 0; i < parameterizedReturnType.arguments().size(); i++) {
+                Type paramReturnTypeArg = parameterizedReturnType.arguments().get(i);
+                Type resolvedType;
+                if (paramReturnTypeArg.kind() == TYPE_VARIABLE) {
+                    // method returns another subresource, and one of the arguments is a type variable e.g.  Wrapper<T>
+                    resolvedType = ownerIdentifierToTypeVariable.get(paramReturnTypeArg.asTypeVariable().identifier());
+                    if (resolvedType == null) {
+                        throw new IllegalArgumentException(
+                                "Type variable %s of the sub resource locator method's return type %s could not be resolved."
+                                        .formatted(paramReturnTypeArg.asTypeVariable().identifier(), jandexMethod));
+                    }
+                } else {
+                    // Subresource, but no type variable, e.g. Wrapper<String>
+                    resolvedType = paramReturnTypeArg;
+                }
+
+                identifierToTypeVariable.put(returnClass.typeParameters().get(i).identifier(), resolvedType);
+                methodReturnTypeBuilder.addArgument(resolvedType);
+            }
+
+            // rewrite returnType to reflect the resolved type variable for the generatedSubResources cache
+            // i.e. Wrapper<String> instead of Wrapper<V>
+            returnType = methodReturnTypeBuilder.build();
+        } else if (returnType.kind() == TYPE_VARIABLE) {
+            TypeVariable typeVariable = returnType.asTypeVariable();
+            // rewrite returnType to reflect the resolved type variable for the generatedSubResources cache
+            // i.e. String instead of Type Variable V
+            returnType = identifierToTypeVariable.get(typeVariable.identifier());
+            if (returnType == null) {
+                return;
+            }
+
+        } else if (returnType.kind() != CLASS) {
             // sort of sub-resource method that returns a thing that isn't a class
             throw new IllegalArgumentException("Sub resource type is not a class: " + returnType.name().toString());
         }
@@ -1854,7 +1895,7 @@ public class JaxrsClientReactiveProcessor {
                     handleReturn(subInterface, defaultMediaType,
                             getHttpMethod(jandexSubMethod, subMethod.getHttpMethod(), httpAnnotationToMethod),
                             consumes, jandexSubMethod, subMethodCreator, formParams, bodyParameterValue,
-                            builder, multipart);
+                            builder, multipart, identifierToTypeVariable);
                 } else {
                     // finding corresponding jandex method, used by enricher (MicroProfile enricher stores it in a field
                     // to later fill in context with corresponding java.lang.reflect.Method)
@@ -1867,7 +1908,7 @@ public class JaxrsClientReactiveProcessor {
                     handleSubResourceMethod(enrichers, generatedClasses, subInterface, index,
                             defaultMediaType, httpAnnotationToMethod, subName, subContext, subMethodTarget,
                             subMethodIndex, subMethod, subJavaMethodParameters, jandexSubMethod,
-                            multipartResponseTypes, subParamFields, generatedSubResources);
+                            multipartResponseTypes, subParamFields, generatedSubResources, identifierToTypeVariable);
                 }
 
             }
@@ -2295,13 +2336,24 @@ public class JaxrsClientReactiveProcessor {
 
     private void handleReturn(ClassInfo restClientInterface, String defaultMediaType, String httpMethod, String[] consumes,
             MethodInfo jandexMethod, MethodCreator methodCreator, ResultHandle formParams,
-            ResultHandle bodyValue, AssignableResultHandle builder, boolean multipart) {
+            ResultHandle bodyValue, AssignableResultHandle builder, boolean multipart,
+            Map<String, Type> identifierToTypeVariable) {
         Type returnType = jandexMethod.returnType();
         ReturnCategory returnCategory = ReturnCategory.BLOCKING;
 
+        if (returnType.kind() == TYPE_VARIABLE) {
+            TypeVariable typeVariable = returnType.asTypeVariable();
+            Type resolvedTypeVariable = identifierToTypeVariable.get(typeVariable.identifier());
+            if (resolvedTypeVariable != null) {
+                returnType = resolvedTypeVariable;
+            } else {
+                throw new RuntimeException("Type variable %s of the return type of method %s could not be resolved."
+                        .formatted(typeVariable.identifier(), jandexMethod));
+            }
+        }
+
         String simpleReturnType = returnType.name().toString();
         ResultHandle genericReturnType = null;
-
         if (returnType.kind() == PARAMETERIZED_TYPE) {
             ParameterizedType paramType = returnType.asParameterizedType();
             if (ASYNC_RETURN_TYPES.contains(paramType.name())) {

--- a/extensions/resteasy-reactive/rest-client/deployment/src/test/java/io/quarkus/rest/client/reactive/subresource/SubResourceGenericsTest.java
+++ b/extensions/resteasy-reactive/rest-client/deployment/src/test/java/io/quarkus/rest/client/reactive/subresource/SubResourceGenericsTest.java
@@ -1,0 +1,208 @@
+package io.quarkus.rest.client.reactive.subresource;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.WebApplicationException;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.MultivaluedMap;
+import jakarta.ws.rs.ext.Provider;
+
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+import org.jboss.resteasy.reactive.client.impl.RestClientRequestContext;
+import org.jboss.resteasy.reactive.client.spi.ClientMessageBodyReader;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.rest.client.reactive.QuarkusRestClientBuilder;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class SubResourceGenericsTest {
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClass(Resource.class)
+                    .addClass(ClientLocator.class)
+                    .addClass(TranslationSubResource.class)
+                    .addClass(EnglishSubResource.class)
+                    .addClass(ShortsSubResource.class)
+                    .addClass(HelloSubResource.class)
+                    .addClass(EachCellAsListElementClientMessageBodyReader.class)
+                    .addClass(CellsAsMapEntryClientMessageBodyReader.class));
+
+    @RestClient
+    ClientLocator client;
+
+    @Test
+    void testRestCalls() {
+        {
+            String result = client.string().subResource().subResource().subResource().get();
+            assertThat(result).isEqualTo("Hello,World");
+        }
+
+        {
+            Long result = client.number().subResource().subResource().subResource().get();
+            assertThat(result).isEqualTo(42L);
+        }
+
+        {
+            List<String> result = client.testListOnRoot();
+            assertThat(result).contains("Hello", "World");
+        }
+
+        {
+            List<String> result = client.list().subResource().subResource().subResource().get();
+            assertThat(result).contains("Hello", "World");
+        }
+
+        {
+            Map<String, String> result = client.map().subResource().subResource().subResource().get();
+            assertThat(result).containsEntry("Hello", "World");
+        }
+    }
+
+    @Test
+    void testFailureSubResourceLocatorMethodWithUnresolvedTypeVariable() {
+
+        try {
+            QuarkusRestClientBuilder.newBuilder().baseUri(URI.create("http://localhost:8081"))
+                    .build(TranslationSubResource.class);
+        } catch (Exception e) {
+            assertThat(e.getMessage()).endsWith(
+                    "Failed to generate client for class interface io.quarkus.rest.client.reactive.subresource.SubResourceGenericsTest$TranslationSubResource : Type variable R of the sub resource locator method's return type io.quarkus.rest.client.reactive.subresource.SubResourceGenericsTest$EnglishSubResource<R> subResource() could not be resolved.");
+            return;
+        }
+        Assertions.fail("Should have thrown an exception");
+    }
+
+    @Test
+    void testFailureRestClientMethodWithUnresolvedTypeVariable() {
+
+        try {
+            QuarkusRestClientBuilder.newBuilder().baseUri(URI.create("http://localhost:8081")).build(HelloSubResource.class);
+        } catch (Exception e) {
+            assertThat(e.getMessage()).endsWith(
+                    "Failed to generate client for class interface io.quarkus.rest.client.reactive.subresource.SubResourceGenericsTest$HelloSubResource : Type variable V of the return type of method V get() could not be resolved.");
+            return;
+        }
+        Assertions.fail("Should have thrown an exception");
+    }
+
+    @Path("")
+    public static class Resource {
+        @Path("greetings/translations/english/shorts/hello")
+        @GET
+        public String greet() {
+            return "Hello,World";
+        }
+
+        @Path("greetings-count/translations/english/shorts/hello")
+        @GET
+        public Long greetingCount() {
+            return 42L;
+        }
+    }
+
+    @RegisterRestClient(baseUri = "http://localhost:8081")
+    @Path("")
+    public interface ClientLocator {
+
+        @Path("greetings/translations")
+        TranslationSubResource<String> string();
+
+        @Path("greetings-count/translations")
+        TranslationSubResource<Long> number();
+
+        @Path("greetings/translations")
+        TranslationSubResource<List<String>> list();
+
+        @Path("greetings/translations")
+        TranslationSubResource<Map<String, String>> map();
+
+        @Path("greetings/translations/english/shorts/hello")
+        @GET
+        List<String> testListOnRoot();
+    }
+
+    public interface TranslationSubResource<R> {
+        @Path("english")
+        EnglishSubResource<R> subResource();
+    }
+
+    public interface EnglishSubResource<Z> {
+        @Path("shorts")
+        ShortsSubResource<Z> subResource();
+    }
+
+    public interface ShortsSubResource<Y> {
+        @Path("hello")
+        HelloSubResource<Y> subResource();
+    }
+
+    public interface HelloSubResource<V> {
+        @GET
+        V get();
+    }
+
+    @Provider
+    public static class EachCellAsListElementClientMessageBodyReader implements ClientMessageBodyReader<List<String>> {
+
+        @Override
+        public List<String> readFrom(Class<List<String>> type, Type genericType, Annotation[] annotations, MediaType mediaType,
+                MultivaluedMap<String, String> httpHeaders, InputStream entityStream, RestClientRequestContext context)
+                throws IOException, WebApplicationException {
+            return readFrom(type, genericType, annotations, mediaType, httpHeaders, entityStream);
+        }
+
+        @Override
+        public boolean isReadable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+            return genericType.getTypeName().equals("java.util.List<java.lang.String>");
+        }
+
+        @Override
+        public List<String> readFrom(Class<List<String>> type, Type genericType, Annotation[] annotations, MediaType mediaType,
+                MultivaluedMap<String, String> httpHeaders, InputStream entityStream)
+                throws IOException, WebApplicationException {
+            String body = new String(entityStream.readAllBytes(), StandardCharsets.UTF_8);
+            return Arrays.asList(body.split(","));
+        }
+    }
+
+    @Provider
+    public static class CellsAsMapEntryClientMessageBodyReader implements ClientMessageBodyReader<Map<String, String>> {
+
+        @Override
+        public Map<String, String> readFrom(Class<Map<String, String>> type, Type genericType, Annotation[] annotations,
+                MediaType mediaType, MultivaluedMap<String, String> httpHeaders, InputStream entityStream,
+                RestClientRequestContext context) throws IOException, WebApplicationException {
+            return readFrom(type, genericType, annotations, mediaType, httpHeaders, entityStream);
+        }
+
+        @Override
+        public boolean isReadable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
+            return genericType.getTypeName().equals("java.util.Map<java.lang.String, java.lang.String>");
+        }
+
+        @Override
+        public Map<String, String> readFrom(Class<Map<String, String>> type, Type genericType, Annotation[] annotations,
+                MediaType mediaType, MultivaluedMap<String, String> httpHeaders, InputStream entityStream)
+                throws IOException, WebApplicationException {
+            String body = new String(entityStream.readAllBytes(), StandardCharsets.UTF_8);
+            String[] split = body.split(",");
+            return Map.of(split[0], split[1]);
+        }
+    }
+}

